operator/identity/crd_gc.go

```diff
@@ -0,0 +1,184 @@
+)
+
+func (igc *GC) startCRDModeGC(ctx context.Context) error {
+	if igc.gcInterval != 0 {
+		igc.logger.WithField(logfields.Interval, igc.gcInterval).Info("Starting CRD identity garbage collector")
+
+		igc.mgr = controller.NewManager()
+		igc.mgr.UpdateController("crd-identity-gc",
+			controller.ControllerParams{
+				RunInterval:  igc.gcInterval,
+				DoFunc:       igc.gc,
+				NoErrorRetry: true,
+			})
+	}
+
+	return igc.wp.Submit("hearbeat-updater", igc.runHeartbeatUpdater)
+}
+
+func (igc *GC) runHeartbeatUpdater(ctx context.Context) error {
+	for event := range igc.identity.Events(ctx) {
+		switch event.Kind {
+		case resource.Upsert:
+			// Identity is marked as alive if it is new or it has
+			// been updated.
+			igc.heartbeatStore.MarkAlive(event.Object.Name, time.Now())
+		case resource.Delete:
+			// When the identity is deleted, delete the
+			// heartbeat entry as well. This will not be
+			// 100% accurate as the CiliumEndpoint can live
+			// longer than the CiliumIdentity. See
+			// heartbeatStore.gc()
+			igc.heartbeatStore.Delete(event.Object.Name)
+		}
+		event.Done(nil)
+	}
+	return nil
+}
+
+// gc is a single iteration of a garbage collection. It will
+// delete identities that have not had its heartbeat lifesign updated
+// since HeartbeatTimeout.
+func (igc *GC) gc(ctx context.Context) error {
+	igc.logger.Debug("Running CRD identity garbage collector")
+
+	select {
+	case <-watchers.CiliumEndpointsSynced:
+	case <-ctx.Done():
+		return nil
+	}
+
+	identities, err := igc.clientset.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		igc.logger.WithError(err).Error("unable to list Cilium identities")
+	}
+	totalEntries := len(identities.Items)
+	deletedEntries := 0
+
+	timeNow := time.Now()
+	for _, identity := range identities.Items {
+		// The identity is definitely alive if there's a CE using it.
+		if watchers.HasCEWithIdentity(identity.Name) {
+			igc.heartbeatStore.MarkAlive(identity.Name, timeNow)
+			continue
+		}
+
+		if !igc.heartbeatStore.IsAlive(identity.Name) {
+			ts, ok := identity.Annotations[identitybackend.HeartBeatAnnotation]
+			if !ok {
+				log.WithFields(logrus.Fields{
+					logfields.Identity: identity.Name,
+					logfields.K8sUID:   identity.UID,
+				}).Info("Marking identity for later deletion")
+
+				if identity.Annotations == nil {
+					identity.Annotations = make(map[string]string)
+				}
+
+				identity.Annotations[identitybackend.HeartBeatAnnotation] = timeNow.Format(time.RFC3339Nano)
+				if err := igc.updateIdentity(ctx, &identity); err != nil {
+					log.WithError(err).
+						WithField(logfields.Identity, identity).
+						Error("Marking identity for later deletion")
+				}
+
+				continue
+			}
+
+			log.WithFields(logrus.Fields{
+				logfields.Identity: identity,
+			}).Debugf("Deleting unused identity; marked for deletion at %s", ts)
+
+			if err := igc.deleteIdentity(ctx, &identity); err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
```

Comment:
Since pkg/controller is invoking this method, it'd be good to collect all errors and return them so we can see the controller as failing. Same also with updateIdentity() on line 98.